docs/codeql/codeql-language-guides/customizing-library-models-for-cpp.rst

```diff
@@ -0,0 +1,184 @@
+
+The extensible predicates are populated using the models defined in data extension files.
+
+Example of custom model definitions
+------------------------------------
+
+The examples in this section are taken from the standard CodeQL CPP query pack published by GitHub. They demonstrate how to add tuples to extend extensible predicates that are used by the standard queries.
+
+Example: Taint source from the ``boost::asio`` namespace
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This example shows how the CPP query pack models the return value from the ``read_until`` function as a ``remote`` source.
+
+.. code-block:: cpp
+
+   boost::asio::read_until(socket, recv_buffer, '\0', error);
+
+We need to add a tuple to the ``sourceModel``\(namespace, type, subtypes, name, signature, ext, output, kind, provenance) extensible predicate by updating a data extension file.
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/cpp-all
+         extensible: sourceModel
+       data:
+         - ["boost::asio", "", False, "read_until", "", "", "Argument[*1]", "remote", "manual"]
+
+Since we are adding a new source, we need to add a tuple to the ``sourceModel`` extensible predicate.
+The first five values identify the callable (in this case a free function) to be modeled as a source.
+
+- The first value ``"boost::asio"`` is the namespace name.
+- The second value ``""`` is the name of the type (class) that contains the method. Because we're modelling a free function, the type is left blank.
+- The third value ``False`` is a flag that indicates whether or not the sink also applies to all overrides of the method. For a free function, this should be ``False``.
+- The fourth value ``"read_until"`` is the function name.
+- The fifth value is the function input type signature, which can be used to narrow down between functions that have the same name. In this case, we want the model to include all functions in ``boost::asio`` called ``read_until``.
+
+The sixth value should be left empty and is out of scope for this documentation.
+The remaining values are used to define the ``access path``, the ``kind``, and the ``provenance`` (origin) of the source.
+
+- The seventh value ``"Argument[*1]"`` is the ``access path``, which means that the sink is the first indirection (or pointed-to value, ``*``) of the second argument (``Argument[1]``) passed to the function.
+- The eighth value ``"remote"`` is the kind of the source. The source kind is used to define the threat model where the source is in scope. ``remote`` applies to many of the security related queries as it means a remote source of untrusted data. For more information, see ":ref:`Threat models <threat-models-cpp>`."
+- The ninth value ``"manual"`` is the provenance of the source, which is used to identify the origin of the source.
```

Comment:
It's really the origin of the source _model_, right? Should we make this more clear in the text?

Reply:
Yeah, again not my words but this definitely sounds like an improvement.

Reply:
Done.